man/xkcdpunk.1
.Dd February 9, 2026
.Dt XKCDPUNK 1
.Os 
.
.Sh NAME
.Nm xkcdpunk
.Nd Display XKCD comics directly in your terminal
.
.Sh SYNOPSIS
.Nm
.Op \-offline
.Op Ar INPUT ...
.Nm
.Fl h | \-help
.
.Sh DESCRIPTION
xkcdpunk will display a given xkcd comic and its alt-text directly in your terminal. Without argument, the latest known comic will be displayed. The argument can also be "random" to display a random comic.
.Ss Positional arguments
.Bl -tag -width Ds -offset indent
.It Ar INPUT
XKCD number. Besides a number, "latest" and "random" are accepted values. If missing, "latest" is assumed
.El
.Ss Keyword arguments
.Bl -tag -width Ds -offset indent
.It Fl \-offline
Only access already cached comics and do not download anything. "latest" will fails if no comic cached ever and "random" might fails if not enough cached comics. 
.El
.
.Sh EXIT STATUS
.Ex -std
.
.Sh SEE ALSO
.Xr ansicat 1 ,
.Xr netcache 1 ,
.Xr offpunk 1 ,
.Xr openk 1 ,
.Lk https://offpunk.net/
.
.Sh AUTHORS
.An Lionel Dricot (Ploum) Aq Mt offpunk2 at ploum.eu


CHANGELOG
# Offpunk History

## 3.1 - Unreleased
- Cut startup time in half by lazy loading optional modules
- "redirect" now has support for "whitelist". Whitelisted websites will always be shown in full view, skipping readability/unmerdify
- Adding support for tmux clipboard and cleaning clipboard code (by Stephen)
- Default ON prompt color set to yellow (by Matin Mohammadi)
- "copy" now default to "url" and works coherently 
- hide pictures which are from blocked URL (and hide IMG links pointing to blocked URL)
- "reply" now also looks for "contact" pages
- CRASHFIX: "list" was crashing because of a bad call to super() (reported by Toby Kurien)
- CRASHFIX: "--features" was crashing (patch by Étienne Mollier)
- BUGFIX: "info" was sometimes returning the wrong cleaning method
- BUGFIX: history latest items was sometimes overwritten with multiple back/forward
- BUGFIX: replace wrongly encoded UTF-8 in Gemini page instead of crashing (Pawel Piatkwoski)
- PACKAGING: use hatch-requirements-txt (by link2xt)
- Lot of typos fixed (by Sebastian Rasmussen)
- Fixed xkcdpunk manpage (by Étienne Mollier)
- "unmerdify" manpage (by Étienne Mollier)
- TRANSLATION: added Swedish SV (by Sebastian Rasmussen)
- TRANSLATION: added Persian (Farsi) (by Mati Mohammadi)
- refactorisation to handle every external binary in one place in offutils.py

## 3.0 - February 9th 2026
Changes since 2.8
Important changes:
- "opnk" is deprecated and replaced by "openk"
- Default image size now 100 instead of 40 (can be restored with "set images_size 40")
- Links to available feeds are now displayed at the bottom of HTML pages
- "root" has been modified to go to the root of the capsule/userspace "root /" for real root
- Images are now displayed in Gemini. This can be disabled with "set gemini_images false"
- "ls" command is deprecated
Features:
- Offpunk is now translatable (by JMCS). Translations in ES, GC (JMCS) and NL (Bert Livens)
- Add "unmerdify" tool to extract content from HTML if "ftr_site_config" is set (by Vincent Jousse)
- new "xkcdpunk" command-line tool to directly access your XKCD comics
- new "websearch" command which default to wiby.me
- new "share" command to send a page by email
- new "reply" command to send an email to the author of the page
- "cookies" command and http cookies are now supported in netcache (by Urja)
- "set default_cmd" now allows to configure what happen on empty lines
- "view switch" changes between readable/full view (by Andrew Fowlie)
- customize prompt with "set prompt_on" and "set prompt_off" (by Andrew Fowlie)
- "help help" now sends an email to the offpunk-users mailing-list
- "links" command now display all links
- new "blocked_link" object in theme. By default, blocked link are in red.
- new "theme preset" to allow switching between multiple hardcoded themes like "yellow","cyan" or "bw"
Others:
- In "lists", only consider files ending with .gmi (bug #46 by woffs)
- support for new links color in gopher (by JMCS)
- fix an off by one bug in "cp url XX" 
- Ansicat: Fix incorrectly identifying RSS feeds as XHTML when there was no entry
- Fix sync hanging on gopher interactive prompt (by JMCS)
- Fix crash when trying to parse HTML without BS4 (reported by Hoël Bézier)
- Automatically replace space by %20 in img url in HTML
- Ansicat: HTML Render img with data-src if src doesn’t link to a valid image
- Opnk: fix a crash when trying to cache-clean uncached url (such as images)
- added mime-type "application/pgp-keys" as to be opened by TextRenderer
- "redirect" has been refactored to acts on the netcache level
- ansicat: in HTML, the <a> link to a picture is only displayed if different from the picture itself
- added "wgl" and "wde" shortcuts to wikipedia GL & DE (JMCS)
- "bugreport" allows to send a bug report to the -devel list
- "theme" now accept "none" as a color to remove part of the theme you don’t want
- "info" now explain how the HTML content was cleaned
- history now consider multiple successive views of a page as one visit only

## 3.0-beta3 - Released as 3.0
- added ftr_site_config output to "version"
- added "wgl" and "wde" shortcuts to wikipedia GL & DE (JMCS)
- TRANSLATION: updated GL and ES (JMCS)
- BUGFIX: successive visits of the same URL are not added to history even in different modes
- TRANSLATION: NL_be (by Bert Livens)
- Added manpage for "xkcdpunk"

## 3.0-beta2 - February 6th 2026
Changes since 3.0-beta1
- "bugreport" allows to send a bug report to the -devel list
- "theme" now accept "none" as a color to remove part of the theme you don’t want
- symlink opnk.py to openk.py to handle deprecation
- BUGFIX: history now consider multiple successive views of a page as one visit.
- BUGFIX: do not urlify mailto, gopher and local urls
- BUGFIX: In "openk", self.last_width was badly reinitialized after cleanup()
- CRASHFIX: in "ansicat", build_body_and_links was crashing with empty links
- CRASHFIX: "openk" was crashing when used in a shell pipe
- improve Gopher selector code to be more robust (by JMCS)
- TRANSLATION: Spanish - ES (by JMCS)
- TRANSLATION: Galician - GC (by JMCS)

## 3.0-beta1 - February 2nd 2026
This is a beta release before 3.0. Its goal is to find bugs, allow for a 
"string freeze" period for translators and give packagers time to adapt 
to the major changes: adding translations to their packages, adding 
xkcdpunk and unmerdify. Packagers are advised to build the package to 
report bugs but *not* to upload to their respective stable 
distribution.
Beta-testers and documenters are welcome to test and document new 
features and new tools.
Important changes:
- "opnk" is deprecated and replaced by "openk"
- Default image size now 100 instead of 40 (can be restored with "set images_size 40")
- Links to available feeds are now displayed at the bottom of HTML pages
- "root" has been modified to go to the root of the capsule/userspace "root /" for real root
- Images are now displayed in Gemini. This can be disabled with "set gemini_images false"

xkcdpunk.py
#!/usr/bin/env python3
import sys
import argparse
import gettext
import random

import netcache
import openk
from offutils import _LOCALE_DIR

gettext.bindtextdomain('offpunk', _LOCALE_DIR)
gettext.textdomain('offpunk')
_ = gettext.gettext

def get_latest(name="xkcd",offline=False):
    if name == "xkcd":
        rss = "https://xkcd.com/atom.xml"
        latest_link = 2
    #validity is 12h = 43200 seconds
    if offline: validity = 0
    else: validity = 43200
    cache = openk.opencache()
    netcache.fetch(rss,validity=validity,offline=offline)
    r = cache.get_renderer(rss)
    if not r:
        print("xkcdpunk needs to be run at least once online to fetch the rss feed")
        return None
    else:
        link = r.get_link(latest_link)
        return link

def main():
    descri = _("xkcdpunk is a tool to display a given XKCD comic in your terminal")
    parser = argparse.ArgumentParser(prog="xkcd", description=descri)
    parser.add_argument(
            "number",
            nargs="*",
            help=_("XKCD comic number. Also accept value \"latest\" and \"random\". Default is \"latest\"")
            )
    parser.add_argument(
            "--offline", action="store_true", help=_("Only access cached comics")
            )
    args = parser.parse_args()
    cache = openk.opencache()
    url = "https://xkcd.com/"
    u = None
    for n in args.number:
        if n.isdigit():
            u = url + str(n) + "/"
        elif n == "random":
            # for the random, we simply take the biggest known xkcd comic
            # and find a number between 1 and latest
            last_url = get_latest(offline=args.offline)
            if last_url:
                last = last_url.strip("/").split("/")[-1]
                if last.isdigit():
                    value = random.randrange(2,int(last))
                    u = url + str(value) + "/"
                    if args.offline:
                        # If offline, we check for a valid netcache version
                        # We introduce a max counter to not search infinitely
                        max_counter = 0
                        while not netcache.is_cache_valid(u) and max_counter < 1000:
                            value = random.randrange(2,int(last))
                            u = url + str(value) + "/"
                            max_counter += 1
        elif n == "latest":
            u = get_latest(offline=args.offline)
    if not u:
        #By default, we get the latest
        u = get_latest(offline=args.offline)
    if u:
        cache.openk(u)
    else:
        print("No cached XKCD comics were found. Please run xkcdpunk online to build the cache")


if __name__ == "__main__":
    main()
